### PR TITLE
Store appended files and deleted files as Content type in IndexLogEntry

### LINKED ISSUE #210: appendedFiles / deletedFiles in IndexLogEntry can be stored as Seq[FileInfo] instead of Seq[String]
> `RefreshDeleteAction` handles the modified files as deleted files, and also as appended files.

Does it make sense to update `appendedFiles` / `deletedFiles` to be `Seq[FileInfo]` in index log entry for completeness (esp. if someone is reading the index log entry)?

_Originally posted by @imback82 in https://github.com/microsoft/hyperspace/pull/207#issuecomment-707868254_

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: Fixes #210 
 - **Parent Issue**: n/a
 - **Dependencies**: n/a

### What changes were proposed in this pull request?

Store the type of appendedFiles and deletedFiles as Content type instead of list of String.
This change is required for #200 and #198.

```
case class Properties(
      content: Content,
-      appendedFiles: Seq[String] = Nil,
-      deletedFiles: Seq[String] = Nil)
+      appendedFiles: Option[Content] = None,
+      deletedFiles: Option[Content] = None)
```

ref) `None` value is serialized as "null" in JSON
```
...
appendedFiles: null
...
```

### Does this PR introduce _any_ user-facing change?

Yes.
This is a breakage change.
Form of `appendedFiles` and `deletedFiles` in IndexLogEntry has changed - a user should rebuild indexes.

### How was this patch tested?

Unit test

### CODE REVIEW

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogEntryTest.scala [hunk 1]
```diff
@@ -126,8 +126,15 @@ class IndexLogEntryTest extends SparkFunSuite with SQLHelper with BeforeAndAfter
         |                  "properties" : { }
         |                }
         |              },
-        |              "deletedFiles" : ["file:/rootpath/f1"],
-        |              "appendedFiles" : ["file:/rootpath/f3"]
+        |              "deletedFiles" : [{
+        |                    "name" : "file:/rootpath/f1",
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/IndexLogEntryTest.scala:130]: Now that I look that this, should we store these as `Directory` to be consistent? The thing that bugs me is that `FileInfo`'s behavior is inconsistent - one is storing full path, and another just name. Also, this can save the metadata file size. Thoughts?

(I know we are using it interchangeably in the code - e.g., `Content.fileInfos`, which can be addressed separately).

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 2]
```diff
@@ -99,17 +99,10 @@ private[actions] abstract class RefreshActionBase(
...
+  protected lazy val deletedFiles: Seq[FileInfo] = {
     val relation = previousIndexLogEntry.relations.head
     val originalFiles = relation.data.properties.content.fileInfos
-    val delFiles = (originalFiles -- currentFiles).map(_.name)
-
-    // Remove duplicate deleted file names in the previous log entry.
-    val prevDeletedFiles = previousIndexLogEntry.deletedFiles.filterNot(delFiles.contains)
-
-    // TODO: Add test for the scenario where existing deletedFiles and newly deleted
-    //  files are updated. https://github.com/microsoft/hyperspace/issues/195.
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:111]: Is this TODO handled in this PR?

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala [hunk 3]
```diff
@@ -389,15 +389,16 @@ case class IndexLogEntry(
...
   }
 
-  def appendedFiles: Seq[String] = {
+  def appendedFiles: Seq[FileInfo] = {
     relations.head.data.properties.appendedFiles
   }
 
-  def withAppendedAndDeletedFiles(appended: Seq[String], deleted: Seq[String]): IndexLogEntry = {
+  def withAppendedAndDeletedFiles(appended: Seq[FileInfo], deleted: Seq[FileInfo]): IndexLogEntry
+  = {
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala:401]: is this by scalafmt? Looks a bit weird. Could you double check?

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 4]
```diff
@@ -99,17 +99,10 @@ private[actions] abstract class RefreshActionBase(
    * Finally, append the previously known deleted files to the result. These
    * are the files for which the index was never updated in the past.
    */
-  protected lazy val deletedFiles: Seq[String] = {
+  protected lazy val deletedFiles: Seq[FileInfo] = {
     val relation = previousIndexLogEntry.relations.head
     val originalFiles = relation.data.properties.content.fileInfos
-    val delFiles = (originalFiles -- currentFiles).map(_.name)
```
Comment (pirz) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:105]: Question: Why is the part for deduplication removed from deletedFiles and appendedFiles functions? (They have been added recently thru #214).

File: src/test/scala/com/microsoft/hyperspace/index/IndexLogEntryTest.scala (hunk 1 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/IndexLogEntryTest.scala:130]: I also thought about using `Content` as utility functions are coupled with it, not with `Directory`. (and inconsistency of FileInfo also bothers me). However, the format below didn't seem good to me, a little. Do you think this is more preferable?

```
case class Hdfs(properties: Hdfs.Properties) {
  val kind = "HDFS"
}
object Hdfs {
  /**
   * Hdfs file properties.
   * @param content Content object representing Hdfs file based data source.
   * @param appendedFiles Appended files since the last time derived dataset was updated.
   * @param deletedFiles Deleted files since the last time derived dataset was updated.
   */
  case class Properties(
      content: Content,
      appendedFiles: Content,
      deletedFiles: Content)
}
```

or with #198
```
case class Update {
  appendedFiles: Content,
  deletedFiles: Content,
  latestSignature: Signature
}
```

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala [hunk 5]
```diff
@@ -334,8 +336,8 @@ object Hdfs {
    */
   case class Properties(
       content: Content,
-      appendedFiles: Seq[String] = Nil,
-      deletedFiles: Seq[String] = Nil)
+      appendedFiles: Content = Content(Directory.emptyDirectory),
+      deletedFiles: Content = Content(Directory.emptyDirectory))
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala:340]: Thanks. I like this approach better. Could you check if we could make it as `Option[Content]`? (If it doesn't work out nicely, `Directory.empty` seems reasonable. (I got `empty` from `def empty[A] : Option[A] = None` for `Option`)

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala (hunk 4 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:105]: Yes, seems we still need to remove duplicates. To my mind, it's a bit of mess to handle duplicates here but #198 will resolve the problem; with "since" semantic, we don't need to care about appendedFiles and deletedFiles of previousIndexLogEntry.

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala (hunk 5 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala]: `Option[Content]` works well. Thanks @imback82!

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala [hunk 6]
```diff
@@ -406,7 +426,8 @@ case class IndexLogEntry(
               relations.head.copy(
                 data = relations.head.data.copy(
                   properties = relations.head.data.properties.copy(
-                    appendedFiles = appended, deletedFiles = deleted))))))))
+                    appendedFiles = Content.fromLeafFiles(appendedFiles),
+                    deletedFiles = Content.fromLeafFiles(deletedFiles)))))))))
   }
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala:432]: How about:
```scala
def toFileStatus(f: FileInfo) = {
  new FileStatus(f.size, false, 0, 1, f.modifiedTime, new Path(f.name))
}
copy(
  source = source.copy(
    plan = source.plan.copy(
      properties = source.plan.properties.copy(
        relations = Seq(
          relations.head.copy(
            data = relations.head.data.copy(
              properties = relations.head.data.properties.copy(
                appendedFiles = Content.fromLeafFiles(appended.map(toFileStatus)),
                deletedFiles = Content.fromLeafFiles(deleted.map(toFileStatus))))))))))
```

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 7]
```diff
@@ -286,11 +286,13 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
       "'appendedFiles' are not usable indexes if hybrid scan is disabled.") {
     withSQLConf(INDEX_HYBRID_SCAN_ENABLED -> "false") {
       val entry1 = createIndexLogEntry("t1iTest", Seq(t1c1), Seq(t1c3), t1ProjectNode)
-      val entry2 = entry1.withAppendedAndDeletedFiles(Seq(), Seq("f1"))
-      val entry3 = entry1.withAppendedAndDeletedFiles(Seq("f2"), Seq())
+      val entry2 = entry1.withAppendedAndDeletedFiles(Seq(), Seq(FileInfo("file:/dir/f1", 1, 1)))
+      val entry3 = entry1.withAppendedAndDeletedFiles(Seq(FileInfo("file:/dir/f2", 1, 1)), Seq())
+      entry2.state = "ACTIVE"
+      entry3.state = "ACTIVE"
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:292]: Is this change required? (It helps reviewers if you make guide comments on changes that are not PR-related).

File: src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala [hunk 8]
```diff
@@ -536,10 +537,11 @@ class RefreshIndexTests extends QueryTest with HyperspaceSuite {
     TestUtils.deleteFiles(dataPath, "*parquet", 1).head
   }
 
-  private def listFiles(path: String): Seq[String] = {
+  private def listFiles(path: String): Seq[FileInfo] = {
     val absolutePath = PathUtils.makeAbsolute(path)
     val fs = absolutePath.getFileSystem(new Configuration)
-    fs.listStatus(absolutePath).toSeq.map(_.getPath.toString)
+    fs.listStatus(absolutePath).toSeq.filter(f => DataPathFilter.accept(f.getPath)).map(f =>
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/RefreshIndexTests.scala:543]: Why is `DataPathFilter` change introduced?

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala [hunk 9]
```diff
@@ -384,20 +390,34 @@ case class IndexLogEntry(
...
+    if (relations.head.data.properties.deletedFiles.isDefined) {
+      relations.head.data.properties.deletedFiles.get.fileInfos
+    } else {
+      Set()
+    }
   }
 
-  def appendedFiles: Seq[String] = {
-    relations.head.data.properties.appendedFiles
+  def appendedFiles: Set[FileInfo] = {
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala:404]: Can you add the following comments to `def`s that return `FileInfo` with full path so we can revisit this later?
```
// FileInfo's 'name' contains the full path to the file.
```

**Review by imback82 — COMMENTED** [review]
Generally looking fine to me.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 10]
```diff
@@ -91,7 +91,7 @@ object RuleUtils {
     }
 
     if (hybridScanEnabled) {
-      // TODO: Duplicate listing files for the given relation as in
+      // TODO: Duplicate listing files for the given 361relation as in
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:94]: typo

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala (hunk 7 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:292]: Oh I see. This is very subtle and I will create an issue to see if we need to follow up (potentially any `copy()` on `IndexLogEntry` could be misleading).

Meanwhile could you add the following?
```
// IndexLogEntry.withAppendedAndDeletedFiles doesn't copy LogEntry's fields. Thus, set the 'state' to ACTIVE manually so that these entries are considered in RuleUtils.getCandidateIndexes.
```

**Review by imback82 — COMMENTED** [review]
LGTM, thanks @sezruby!

**Review by imback82 — APPROVED** [review]

**Review by pirz — APPROVED** [review]
LGTM, thanks @sezruby
